### PR TITLE
feat(secretsmanager): adds grantWrite to Secret

### PR DESCRIPTION
**Pull request by ddneilson**
### Commit Message
feat(secretsmanager): adds grantWrite to Secret
### End Commit Message

Implements: https://github.com/aws/aws-cdk/issues/7682

#### Testing

I deployed the following:
```
#!/usr/bin/env node

import * as cdk from '@aws-cdk/core';
import * as kms from '@aws-cdk/aws-kms';
import * as secretsmanager from '@aws-cdk/aws-secretsmanager';
import * as ec2 from '@aws-cdk/aws-ec2';

const app = new cdk.App();

const stack = new cdk.Stack(app, 'Stack', {});
const key = new kms.Key(stack, 'KMS');
const secret = new secretsmanager.Secret(stack, 'Secret', { encryptionKey: key });
const secret2 = new secretsmanager.Secret(stack, 'Secret2', {});

const vpc = ec2.Vpc.fromVpcAttributes(stack, 'Vpc', {
    vpcId: "vpc-XXXX",
    availabilityZones: [ "us-west-2a" ],
    publicSubnetIds: [ "subnet-XXXX" ],
});
const instance = new ec2.BastionHostLinux(stack, 'Bastion', { 
    vpc: vpc
});
secret.grantRead(instance);
secret.grantWrite(instance);
secret2.grantRead(instance);
secret2.grantWrite(instance);

app.synth();
```

Then, once the stack is deployed I used ssm to connect to the bastion and ran:
```
for secret in <secret name 1> <secret name 2>
do
	aws --region us-west-2 secretsmanager put-secret-value --secret-id ${secret} --secret-string "FooValue"
	aws --region us-west-2 secretsmanager get-secret-value --secret-id ${secret}
done
```

Finally, I verified the contents of the two secrets in the SecretsManager console.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by ddneilson** [comment]
@skinny85 Ready for review whenever you are.

File: packages/@aws-cdk/aws-secretsmanager/README.md [hunk 1]
```diff
@@ -39,6 +39,30 @@ const secret = secretsmanager.Secret.fromSecretAttributes(scope, 'ImportedSecret
...
 
+### Grant permission to use the secret to a role
+
+You must grant permission to a resource for that resource to be allowed to 
+use a secret. This can be achieved with the `Secret.grantRead` and/or 
+`Secret.grantWrite` method, depending on your need:
+
+```ts
+const role = new iam.Role(stack, 'SomeRole', { assumedBy: new iam.AccountRootPrincipal() });
+const secret = new secretsmanager.Secret(stack, 'Secret', {});
```
Comment (skinny85) [inline packages/@aws-cdk/aws-secretsmanager/README.md:50]: You can leave out the `props` argument, it's optional for `Secret`. So, just `new secretsmanager.Secret(stack, 'Secret');`.

File: packages/@aws-cdk/aws-secretsmanager/README.md [hunk 2]
```diff
@@ -39,6 +39,30 @@ const secret = secretsmanager.Secret.fromSecretAttributes(scope, 'ImportedSecret
...
+secret.grantRead(role);
+secret.grantWrite(role);
+```
+
+If, as in the following example, your secret was created with a KMS key
```
Comment (skinny85) [inline packages/@aws-cdk/aws-secretsmanager/README.md:55]: `:` at the end, os `with a KMS key:`.

File: packages/@aws-cdk/aws-secretsmanager/README.md [hunk 3]
```diff
@@ -39,6 +39,30 @@ const secret = secretsmanager.Secret.fromSecretAttributes(scope, 'ImportedSecret
...
+```ts
+const key = new kms.Key(stack, 'KMS');
+const secret = new secretsmanager.Secret(stack, 'Secret', { encryptionKey: key });
+secret.grantRead(role);
+secret.grantWrite(role);
+```
+then `Secret.grantRead` and `Secret.grantWrite` will also grant the role the
+relevant encrypt and decrypt permissions to the KMS key through the
+SecretsManager service principle.
```
Comment (skinny85) [inline packages/@aws-cdk/aws-secretsmanager/README.md:64]: principle -> principal :)

File: packages/@aws-cdk/aws-secretsmanager/lib/secret.ts [hunk 4]
```diff
@@ -148,6 +155,25 @@ abstract class SecretBase extends Resource implements ISecret {
...
+    // @see https://docs.aws.amazon.com/secretsmanager/latest/userguide/auth-and-access_identity-based-policies.html
+    const result = iam.Grant.addToPrincipal({
+      grantee,
+      actions: ['secretsmanager:PutSecretValue'],
+      resourceArns: [this.secretArn],
+      scope: this,
+    });
+
+    if (this.encryptionKey) {
+      // @see https://docs.aws.amazon.com/kms/latest/developerguide/services-secrets-manager.html
```
Comment (skinny85) [inline packages/@aws-cdk/aws-secretsmanager/lib/secret.ts:168]: `@see` is only for JSDocs, doesn't do anything for inline docs, so feel free to remove it :)

File: packages/@aws-cdk/aws-secretsmanager/test/test.secret.ts [hunk 5]
```diff
@@ -435,6 +435,215 @@ export = {
...
+      PolicyDocument: {
+        Version: '2012-10-17',
+        Statement: [{
+          Action: 'secretsmanager:PutSecretValue',
+          Effect: 'Allow',
+          Resource: { Ref: 'SecretA720EF05' },
+        }],
+      },
+    }));
+    expectStack.to(haveResource('AWS::KMS::Key', {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-secretsmanager/test/test.secret.ts:483]: Can we use `haveResourceLike` instead, to match only part of the resource definition here? (Only the one that's relevant to this test)

File: packages/@aws-cdk/aws-secretsmanager/README.md (hunk 1 above)
Comment (ddneilson) [reply packages/@aws-cdk/aws-secretsmanager/README.md]: fixed

File: packages/@aws-cdk/aws-secretsmanager/README.md (hunk 2 above)
Comment (ddneilson) [reply packages/@aws-cdk/aws-secretsmanager/README.md]: fixed

File: packages/@aws-cdk/aws-secretsmanager/README.md (hunk 3 above)
Comment (ddneilson) [reply packages/@aws-cdk/aws-secretsmanager/README.md]: heh. fixed

File: packages/@aws-cdk/aws-secretsmanager/lib/secret.ts (hunk 4 above)
Comment (ddneilson) [reply packages/@aws-cdk/aws-secretsmanager/lib/secret.ts]: That's what I get for blindly copying `grantRead` :-)

File: packages/@aws-cdk/aws-secretsmanager/test/test.secret.ts (hunk 5 above)
Comment (ddneilson) [reply packages/@aws-cdk/aws-secretsmanager/test/test.secret.ts:483]: Turns out the answer is 'no, we can't, unless there's some trick that I don't know for arrays of IAM statements.'

The implementation for `haveResourceLike` will always reject arrays of different length ( https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/assert/lib/assertions/have-resource.ts#L140 ), and the differing parts here are the list/array of IAM policy statements.

Fixing/enhancing `HaveResourceAssertion` to be able to figure out subsets of IAM policy statements would be super useful, but definitely out-of-scope for this PR; it's also likely to have larger ramifications that would require some thought.

It's a shame too, there's a tonne of duplication of the KMS policy in the secrets tests that would be nice to de-duplicate.

**Comment by ddneilson** [comment]
@skinny85 Ready for round 2 when you are.

File: packages/@aws-cdk/aws-secretsmanager/test/test.secret.ts (hunk 5 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-secretsmanager/test/test.secret.ts:483]: That is true, it's not ideal, but it allows you to write:

```ts
    expectStack.to(haveResourceLike('AWS::KMS::Key', {
      KeyPolicy: {
        Statement: [
          {},
          {},
          {},
          {
            Action: [
              'kms:Encrypt',
              'kms:ReEncrypt*',
              'kms:GenerateDataKey*',
            ],
            Condition: {
              StringEquals: {
                'kms:ViaService': {
                  'Fn::Join': ['', [
                    'secretsmanager.',
                    { Ref: 'AWS::Region' },
                    '.amazonaws.com',
                  ]],
                },
              },
            },
            Principal: {
              AWS: {
                'Fn::GetAtt': [
                  'Role1ABCC5F0',
                  'Arn',
                ],
              },
            },
            Resource: '*',
          },
        ],
      },
    });
```

Which is quite a bit more readable.

Comment (ddneilson) [reply packages/@aws-cdk/aws-secretsmanager/test/test.secret.ts:483]: Ahhh. So, there *is* a trick. :-)

I'll apply that, and get a new version up later today. I'll probably also clean up the other tests in the file while I'm at it...

Comment (ddneilson) [reply packages/@aws-cdk/aws-secretsmanager/test/test.secret.ts]: @skinny85 Done. I cleaned up the tests for `grantRead` to use `haveResourceLike` as well.

**Pull request merged**